components/login.js

```diff
@@ -0,0 +1,259 @@
+
+			let {new_client_id: newClientId, refresh_token: refreshToken, access_token: accessToken} = pollRes;
+			// This behavior is a guess and is unconfirmed
+			if (newClientId) {
+				this.emit('debug', `login: got new client id ${newClientId}`);
+				clientId = newClientId;
+			}
+
+			if (refreshToken) {
+				// Access token doesn't seem to be used for anything that I can see right now
+				let finalizeResponse = await new Promise((resolve, reject) => {
+					this.emit('debug', 'login: finalizing login using refresh token');
+					this.httpRequestPost({
+						uri: 'https://login.steampowered.com/jwt/finalizelogin',
+						form: {
+							nonce: refreshToken,
+							sessionid: this.getSessionID(), // the cookie won't actually get sent since this is a different domain, but that matches official behavior
+							redir: 'https://steamcommunity.com/login/home/?goto='
+						},
+						headers: API_HEADERS,
+						json: true
+					}, (err, res, body) => {
+						if (err) {
+							return reject(err);
+						}
+
+						if (res.statusCode != 200) {
+							return reject(new Error('HTTP error ' + res.statusCode));
+						}
+
+						if (!body.transfer_info) {
+							return reject(new Error('Malformed login response'));
+						}
+
+						return resolve(body);
+					});
+				});
+
+				let transfers = finalizeResponse.transfer_info.map(({url, params}) => new Promise((resolve, reject) => {
+					this.emit('debug', `login: transferring login result using url ${url}`);
+					this.httpRequestPost({
+						uri: url,
+						form: {
+							steamID: steamid,
+							...params
+						}
+					}, (err, res) => {
+						if (err) {
+							return reject(err);
+						}
+
+						if (!res.headers || !res.headers['set-cookie']) {
+							return reject(new Error('No Set-Cookie header in result'));
+						}
+
+						let loginCookie = res.headers['set-cookie'].find(c => c.startsWith('steamLoginSecure='));
+						if (!loginCookie) {
+							return reject(new Error('No steamLoginSecure cookie in result'));
+						}
+
+						resolve(loginCookie.split(';')[0].trim());
+					});
+				}));
+
+				let txResult = await Promise.any(transfers);
+				this.setCookies([txResult]); // this should already be set, but just make sure. this also sets our steamID property
+				return {cookies: [txResult], steamID: steamid};
+			}
+
+			// No refresh token received
+			await new Promise(resolve => setTimeout(resolve, interval * 1000));
+		}
+
+		// Polling timed out
+		throw new Error('Login attempt timed out');
+	};
+
+	exec().then(result => callback(null, result)).catch(err => callback(err));
+}
+
+SteamCommunity.prototype._loginApiReq = function(httpMethod, apiMethod, inputData) {
+	let httpReqOptions = {
+		method: httpMethod,
+		uri: `https://api.steampowered.com/IAuthenticationService/${apiMethod}/v1`,
+		headers: API_HEADERS,
+		encoding: null
+	};
+
+	if (inputData) {
+		let proto = Protos[`CAuthentication_${apiMethod}_Request`];
+		let inputBuffer = proto.encode(inputData).finish();
```

Comment:
You don't have to use `input_protobuf_encoded`, web apis accept plain old form submits just fine, and return json data.

Reply:
I'm aware. The intent is to mimic official behavior as closely as possible.

Reply:
Interesting. I do think it's a good idea to at least write a comment, so people looking at these functions will know about it.